### PR TITLE
Online expansion support

### PR DESCRIPTION
**Pull request by shuo-wu**
Some of the changes are actually from [the deprecated PR](https://github.com/longhorn/go-iscsi-helper/pull/17)

longhorn/longhorn#1674

### CODE REVIEW

File: iscsi/initiator.go [hunk 1]
```diff
@@ -348,3 +348,18 @@ func CleanupScsiNodes(target string, ne *util.NamespaceExecutor) error {
...
+		"-T", target,
+		"-R",
+	}
+	if ip != "" {
+		opts = append(opts, "-p", ip)
+	}
+	if _, err := ne.Execute(iscsiBinary, opts); err != nil {
+		return err
+	}
+	return nil
```
Comment (derekbit) [inline iscsi/initiator.go:364]:
```suggestion
	_, err := ne.Execute(iscsiBinary, opts)
	return err
```

File: longhorndev/dev.go [hunk 2]
```diff
@@ -404,16 +404,45 @@ func (d *LonghornDevice) Expand(size int64) error {
...
+		return nil
+	}
+	if err := d.scsiDevice.UpdateScsiBackingStore("longhorn", fmt.Sprintf("size=%v", d.size)); err != nil {
+		return err
+	}
+
+	switch d.frontend {
+	case types.FrontendTGTBlockDev:
+		logrus.Infof("device %v: Expanding frontend %v target %v", d.name, d.frontend, d.scsiDevice.Target)
+		if err := d.scsiDevice.ExpandTarget(size); err != nil {
```
Comment (derekbit) [inline longhorndev/dev.go:424]: Do we need a lock protection during expansion?

Comment (shuo-wu) [reply longhorndev/dev.go:424]: There is already a lock at the beginning of this function. Do we need anything else?

File: iscsidev/iscsi.go [hunk 3]
```diff
@@ -180,6 +181,29 @@ func (dev *Device) StopInitiator() error {
 	return nil
 }
 
+func (dev *Device) RefreshInitiator() error {
+	lock := nsfilelock.NewLockWithTimeout(util.GetHostNamespacePath(HostProc), LockFile, LockTimeout)
+	if err := lock.Lock(); err != nil {
+		return fmt.Errorf("Fail to lock: %v", err)
```
Comment (innobead) [inline iscsidev/iscsi.go:187]: `return errors.Wrap(err, "failed to lock")``

Could you also fix other errors here like `errors.Wrapf(err, "failed to lock")` -> `errors.Wrap(err, "failed to lock")`

File: longhorndev/dev.go [hunk 4]
```diff
@@ -404,16 +404,45 @@ func (d *LonghornDevice) Expand(size int64) error {
...
-
 	if d.size > size {
 		return fmt.Errorf("device %v: cannot expand the device from size %v to a smaller size %v", d.name, d.size, size)
 	} else if d.size == size {
 		return nil
 	}
 	d.size = size
 
+	if d.scsiDevice == nil {
+		return nil
```
Comment (innobead) [inline longhorndev/dev.go:415]: Suggest returning error, because this is not a valid case.

File: longhorndev/dev.go [hunk 5]
```diff
@@ -404,16 +404,45 @@ func (d *LonghornDevice) Expand(size int64) error {
...
+		if err := d.scsiDevice.ExpandTarget(size); err != nil {
+			return fmt.Errorf("device %v: fail to expand target %v: %v", d.name, d.scsiDevice.Target, err)
+		}
+		logrus.Infof("device %v: Expanded frontend %v size to %d, users need to refresh/rescan the initiator by themselves", d.name, d.frontend, size)
+		break
+	case "":
+		logrus.Infof("device %v: skip expansion since the frontend not enabled", d.name)
+		break
+	default:
+		return fmt.Errorf("device %v: unknown frontend %v for expansion", d.name, d.frontend)
```
Comment (innobead) [inline longhorndev/dev.go:444]: `"failed to expand device %v: unknown frontend %v"`

**Review by innobead — CHANGES_REQUESTED** [review]
LGTM, except a few comments.

File: longhorndev/dev.go (hunk 4 above)
Comment (shuo-wu) [reply longhorndev/dev.go:415]: Nope... This means the frontend is disabled. It's still a valid case and the expansion function doesn't need to do anything for it.

Comment (shuo-wu) [reply longhorndev/dev.go:415]: But I will add a info log anyway.

Comment (innobead) [reply longhorndev/dev.go:415]: yeah, it's fair.

**Review by innobead — APPROVED** [review]

File: longhorndev/dev.go (hunk 2 above)
Comment (derekbit) [reply longhorndev/dev.go:425]: Ah, I see. It is fine.

**Pull request merged**